### ISSUE TITLE
Feature request: Modifying messages retains a history of modifications

### ISSUE THREAD

**Issue by eric-vlaanderen** [dwp/queue-triage#87]
Would be good to have the ability to "undo" message content changes when you royally screw things up.

**Comment by amcghie** [comment]
Currently a FailedMessage in the database looks like this:
```javascript
{
    "_id" : "28695988-ad74-4957-b1a1-65ec31a0641d",
    "jmsMessageId" : "ID:localhost.localdomain-35141-1520333891427-10:1:1:1:1",
    "destination" : {
      "brokerName" : "internal-broker",
      "name" : "some-queue"
    },
    "content" : "elixir",
    "properties" : "{\"@class\":\"java.util.HashMap\",\"dlqDeliveryFailureCause\":\"java.lang.Throwable: Delivery[1] exceeds redelivery policy limit:RedeliveryPolicy {destination = null, collisionAvoidanceFactor = 0.15, maximumRedeliveries = 0, maximumRedeliveryDelay = -1, initialRedeliveryDelay = 1000, useCollisionAvoidance = false, useExponentialBackOff = false, backOffMultiplier = 5.0, redeliveryDelay = 1000, preDispatchCheck = true}, cause:null\"}",
    "statusHistory" : [
        {
            "status" : "SENT",
            "effectiveDateTime" : ISODate("2018-03-06T10:43:43.804Z"),
            "_lastModifiedDateTime" : ISODate("2018-03-06T10:43:43.805Z")
        },
        {
            "status" : "RESEND",
            "effectiveDateTime" : ISODate("2018-03-06T10:43:42.609Z"),
            "_lastModifiedDateTime" : ISODate("2018-03-06T10:43:42.610Z")
        },
        {
            "status" : "FAILED",
            "effectiveDateTime" : ISODate("2018-03-06T10:43:42.575Z"),
            "_lastModifiedDateTime" : ISODate("2018-03-06T10:43:42.575Z")
        }
    ],
    "labels" : [ "PR-1234" ]
}
```

If a message that has been processed, classified and resent by the Queue Triage application Dead Letters again, my original thinking was that we would add a new "FAILED" statusHistory entry (these are always added to the top of the array) and update the following fields with the latest values:
- jmsMessageId
- destination
- content
- properties

The drawback of that approach is that we lose the history especially if during Message Classification (or via the UI) the message content or properties are changed.  One approach would be to store these properties with each statusHistory change, so the structure would look like this:

```javascript
{
    "_id" : "28695988-ad74-4957-b1a1-65ec31a0641d",
    "statusHistory" : [
        {
            "status" : "SENT",
            "effectiveDateTime" : ISODate("2018-03-06T10:43:43.804Z"),
            "_lastModifiedDateTime" : ISODate("2018-03-06T10:43:43.805Z")
            "jmsMessageId" : "ID:localhost.localdomain-35141-1520333891427-10:1:1:1:1",
            "destination" : {
                "brokerName" : "internal-broker",
                "name" : "another-queue"
            },
            "content" : "{\"foo\": \"bar\"}",
            "properties" : "{\"@class\":\"java.util.HashMap\",\"failedMessageId\":\"28695988-ad74-4957-b1a1-65ec31a0641d\"}", <-- NOTE: failedMessageId property is automatically set on Sending
        },
        {
            "status" : "RESEND",
            "effectiveDateTime" : ISODate("2018-03-06T10:43:42.609Z"),
            "_lastModifiedDateTime" : ISODate("2018-03-06T10:43:42.610Z")
            "jmsMessageId" : "ID:localhost.localdomain-35141-1520333891427-10:1:1:1:1",
            "destination" : {
                "brokerName" : "internal-broker",
                "name" : "another-queue"                         <-- NOTE: Sending to a new Queue
            },
            "content" : "{\"foo\": \"bar\"}",                    <-- NOTE: Content has changed
            "properties" : "{\"@class\":\"java.util.HashMap\"}", <-- NOTE: dlqDeliveryFailureCause property is automatically removed by Queue Triage
        },
        {
            "status" : "FAILED",
            "effectiveDateTime" : ISODate("2018-03-06T10:43:42.575Z"),
            "jmsMessageId" : "ID:localhost.localdomain-35141-1520333891427-10:1:1:1:1",
            "destination" : {
                "brokerName" : "internal-broker",
                "name" : "some-queue"
            },
            "content" : "elixir",
            "properties" : "{\"@class\":\"java.util.HashMap\",\"dlqDeliveryFailureCause\":\"java.lang.Throwable: Delivery[1] exceeds redelivery policy limit:RedeliveryPolicy {destination = null, collisionAvoidanceFactor = 0.15, maximumRedeliveries = 0, maximumRedeliveryDelay = -1, initialRedeliveryDelay = 1000, useCollisionAvoidance = false, useExponentialBackOff = false, backOffMultiplier = 5.0, redeliveryDelay = 1000, preDispatchCheck = true}, cause:null\"}",
        }
    ],
    "labels" : [ "PR-1234" ]
}
```

This change shouldn't be too intrusive as the majority of the time we would only be interested in the latest statusHistory item.  An alternative would be to store the latest "state" at the document root and add to the statusHistory array when a state change occurs.

**Comment by amcghie** [comment]
Resolved by #165
